auth/ante_test.go

```diff
@@ -0,0 +1,367 @@
+
+	// Check no signatures fails
+	checkInvalidTx(t, anteHandler, ctx, tx1, false, sdk.CodeNoSignatures)
+
+	// unknown address error
+	msg2 := sdkAuth.NewTestMsg(addr1) // using first address
+	tx2 := types.NewTestTx(ctx, msg2, priv2, uint64(0), uint64(0))
+
+	// Check no signatures fails
+	checkInvalidTx(t, anteHandler, ctx, tx2, false, sdk.CodeUnknownAddress)
+}
+
+func (suite *AnteTestSuite) TestAccountNumbers() {
+	t, happ, ctx, anteHandler := suite.T(), suite.app, suite.ctx, suite.anteHandler
+
+	// keys and addresses
+	priv1, _, addr1 := sdkAuth.KeyTestPubAddr()
+	_, _, addr2 := sdkAuth.KeyTestPubAddr()
+
+	// set the accounts
+	acc1 := happ.AccountKeeper.NewAccountWithAddress(ctx, hmTypes.AccAddressToHeimdallAddress(addr1))
+	acc1.SetCoins(simulation.RandomFeeCoins())
+	require.NoError(t, acc1.SetAccountNumber(0))
+	happ.AccountKeeper.SetAccount(ctx, acc1)
+	acc2 := happ.AccountKeeper.NewAccountWithAddress(ctx, hmTypes.AccAddressToHeimdallAddress(addr2))
+	acc2.SetCoins(simulation.RandomFeeCoins())
+	require.NoError(t, acc2.SetAccountNumber(1))
+	happ.AccountKeeper.SetAccount(ctx, acc2)
+
+	// msg and signatures
+	var tx sdk.Tx
+	msg := sdkAuth.NewTestMsg(addr1)
+
+	// test good tx from one signer
+	tx = types.NewTestTx(ctx, msg, priv1, uint64(0), uint64(0))
+	checkValidTx(t, anteHandler, ctx, tx, false)
+
+	// new tx from wrong account number
+	tx = types.NewTestTx(ctx, msg, priv1, uint64(1), uint64(0))
+	checkInvalidTx(t, anteHandler, ctx, tx, false, sdk.CodeUnauthorized)
+
+	// from correct account number
+	tx = types.NewTestTx(ctx, msg, priv1, uint64(0), uint64(1))
+	checkValidTx(t, anteHandler, ctx, tx, false)
+}
+
+// Test logic around account number checking with many signers when BlockHeight is 0.
+func (suite *AnteTestSuite) TestAccountNumbersAtBlockHeightZero() {
+	t, happ, ctx, anteHandler := suite.T(), suite.app, suite.ctx, suite.anteHandler
+
+	// keys and addresses
+	priv1, _, addr1 := sdkAuth.KeyTestPubAddr()
+	priv2, _, addr2 := sdkAuth.KeyTestPubAddr()
+
+	// set the accounts, we don't need the acc numbers as it is in the genesis block
+	acc1 := happ.AccountKeeper.NewAccountWithAddress(ctx, hmTypes.AccAddressToHeimdallAddress(addr1))
+	acc1.SetCoins(simulation.RandomFeeCoins())
+	happ.AccountKeeper.SetAccount(ctx, acc1)
+	acc2 := happ.AccountKeeper.NewAccountWithAddress(ctx, hmTypes.AccAddressToHeimdallAddress(addr2))
+	acc2.SetCoins(simulation.RandomFeeCoins())
+	require.NoError(t, acc2.SetAccountNumber(100))
+	happ.AccountKeeper.SetAccount(ctx, acc2)
+
+	// msg and signatures
+	var tx sdk.Tx
+	msg1 := sdkAuth.NewTestMsg(addr1)
+	msg2 := sdkAuth.NewTestMsg(addr2)
+
+	acc1 = happ.AccountKeeper.GetAccount(ctx, hmTypes.AccAddressToHeimdallAddress(addr1))
+	acc2 = happ.AccountKeeper.GetAccount(ctx, hmTypes.AccAddressToHeimdallAddress(addr2))
+	// accNumber1 := acc1.GetAccountNumber()
+	// accNumber2 := acc2.GetAccountNumber()
+
+	// test good tx from one signer
```

Comment:
Should we use a loop for these tests?
```
create slice of struct of test cases
run them via loop
pass them to required functions
```

Reply:
Yeah, better suggestion. We should do that. We can use `t.Run` as well.